### PR TITLE
params: core: enable shanghai based on timestamps

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This PR changes Shanghai activation to be based on timestamps instead of block numbers

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -567,9 +567,9 @@ func (c *ChainConfig) IsTerminalPoWBlock(parentTotalDiff *big.Int, totalDiff *bi
 	return parentTotalDiff.Cmp(c.TerminalTotalDifficulty) < 0 && totalDiff.Cmp(c.TerminalTotalDifficulty) >= 0
 }
 
-// IsShanghai returns whether num is either equal to the Shanghai fork block or greater.
-func (c *ChainConfig) IsShanghai(num *big.Int) bool {
-	return isForked(c.ShanghaiBlock, num)
+// IsShanghai returns whether time is either equal to the Shanghai fork time or greater.
+func (c *ChainConfig) IsShanghai(time *big.Int) bool {
```
Comment (lightclient) [inline params/config.go:571]:
```suggestion
func (c *ChainConfig) IsShanghai(time uint64) bool {
```
It would be nice if either this took a `uint64` or if we don't want `IsShanghai` to take a different parameter  type than other `isForked` methods, maybe a new getter on `types.Block` and `types.Header` "`TimeBig()`" so that each time we check `IsShanghai`, don't need to jump through a bunch of casting hoops?

Comment (roberto-bayardo) [reply params/config.go:571]: It might be better to have a different type here to detect at compile time if someone is trying to pass in a block number instead of a timestamp.  e.g. I see at least one such call in core/chain_makers.go that should be updated.

File: core/blockchain.go [hunk 2]
```diff
@@ -694,13 +713,18 @@ func (bc *BlockChain) setHeadBeyondRoot(head uint64, root common.Hash, repair bo
...
 		if target, force := updateFn(bc.db, bc.CurrentBlock().Header()); force {
-			bc.hc.SetHead(target, updateFn, delFn)
+			bc.hc.SetHead(target.Number.Uint64(), updateFn, delFn)
 		}
 	} else {
 		// Rewind the chain to the requested head and keep going backwards until a
 		// block with a state is found or fast sync pivot is passed
-		log.Warn("Rewinding blockchain", "target", head)
-		bc.hc.SetHead(head, updateFn, delFn)
+		if head != 0 || time == 0 {
```
Comment (MariusVanDerWijden) [inline core/blockchain.go:721]: I'm wondering if it would be better to turn this around (like in `NewBlockchain`)
So setHeadWithTimestamp if time != 0 and setHead otherwise. So a sethead to the genesis block will still be done by number

**Comment by MariusVanDerWijden** [comment]
Forkid filtering is broken in this pr, needs https://github.com/MariusVanDerWijden/go-ethereum/pull/42

File: core/headerchain.go [hunk 3]
```diff
@@ -591,9 +609,9 @@ func (hc *HeaderChain) SetHead(head uint64, updateFn UpdateHeadBlocksCallback, d
 		markerBatch := hc.chainDb.NewBatch()
 		if updateFn != nil {
 			newHead, force := updateFn(markerBatch, parent)
-			if force && newHead < head {
-				log.Warn("Force rewinding till ancient limit", "head", newHead)
-				head = newHead
+			if force && ((headTime > 0 && newHead.Time < headTime) || (headTime == 0 && newHead.Number.Uint64() < headBlock)) {
```
Comment (MariusVanDerWijden) [inline core/headerchain.go:612]: Isn't this just the same condition as `done(newHead)`? It passes the tests when I replace it with that

File: core/blockchain.go (hunk 2 above)
Comment (karalabe) [reply core/blockchain.go:721]: Ah sure, that works

File: core/headerchain.go (hunk 3 above)
Comment (karalabe) [reply core/headerchain.go:612]: Almost. Done stops when newHead.Number <= headBlock, this one only cares if it's <, if it's ==, then no more rewind is done.

**Review by karalabe — APPROVED** [review]
SGTM, but I made a ton of changes, so would appreciate a review from someone else now.

**Review by MariusVanDerWijden — COMMENTED** [review]
LGTM, reviewed @karalabe's latest changes

**Review by lightclient — COMMENTED** [review]
Took a look at the new changes and tests, generally all seem good.

File: core/headerchain.go [hunk 4]
```diff
@@ -566,15 +566,33 @@ type (
...
+// Everything above the new head will be deleted and the new one set.
+func (hc *HeaderChain) setHead(headBlock uint64, headTime uint64, updateFn UpdateHeadBlocksCallback, delFn DeleteBlockContentCallback) {
 	var (
 		parentHash common.Hash
 		batch      = hc.chainDb.NewBatch()
 		origin     = true
 	)
-	for hdr := hc.CurrentHeader(); hdr != nil && hdr.Number.Uint64() > head; hdr = hc.CurrentHeader() {
+	done := func(header *types.Header) bool {
+		if headBlock != 0 || headTime == 0 {
```
Comment (rjl493456442) [inline core/headerchain.go:587]: Maybe we can simplify it by 

```
if headTime == 0 {
    return header.Number.Uint64() <= headBlock
}
return header.Time <= headTime
```

If headTime is zero, it must be block mode;
If headTime is non-zero, it must be time mode;

File: core/headerchain.go [hunk 5]
```diff
@@ -591,9 +609,9 @@ func (hc *HeaderChain) SetHead(head uint64, updateFn UpdateHeadBlocksCallback, d
 		markerBatch := hc.chainDb.NewBatch()
 		if updateFn != nil {
 			newHead, force := updateFn(markerBatch, parent)
-			if force && newHead < head {
-				log.Warn("Force rewinding till ancient limit", "head", newHead)
-				head = newHead
+			if force && ((headTime > 0 && newHead.Time < headTime) || (headTime == 0 && newHead.Number.Uint64() < headBlock)) {
+				log.Warn("Force rewinding till ancient limit", "head", newHead.Number.Uint64())
+				headBlock, headTime = newHead.Number.Uint64(), 0
```
Comment (rjl493456442) [inline core/headerchain.go:614]: Perhaps

`headBlock, headTime = newHead.Number.Uint64(), newHead.Time`
to make two conditions aligned?

File: params/config.go [hunk 6]
```diff
@@ -371,9 +371,12 @@ type ChainConfig struct {
 	ArrowGlacierBlock   *big.Int `json:"arrowGlacierBlock,omitempty"`   // Eip-4345 (bomb delay) switch block (nil = no fork, 0 = already activated)
 	GrayGlacierBlock    *big.Int `json:"grayGlacierBlock,omitempty"`    // Eip-5133 (bomb delay) switch block (nil = no fork, 0 = already activated)
 	MergeNetsplitBlock  *big.Int `json:"mergeNetsplitBlock,omitempty"`  // Virtual fork after The Merge to use as a network splitter
-	ShanghaiBlock       *big.Int `json:"shanghaiBlock,omitempty"`       // Shanghai switch block (nil = no fork, 0 = already on shanghai)
 	CancunBlock         *big.Int `json:"cancunBlock,omitempty"`         // Cancun switch block (nil = no fork, 0 = already on cancun)
```
Comment (rjl493456442) [inline params/config.go:374]: I feel very weird that we still define `CancunBlock` and I am not sure if it's safe or not.
Cancun fork should be the fork after Shanghai, and it must be timestamp based fork.
In forkID, we hold the assumption that we go through all block-based forks first, then
timestamp-based ones.

Can we also define it as `CancunTime` after `ShanghaiTime` and leave it as nil?

File: params/config.go [hunk 7]
```diff
@@ -620,91 +629,107 @@ func (c *ChainConfig) CheckConfigForkOrder() error {
 		{name: "arrowGlacierBlock", block: c.ArrowGlacierBlock, optional: true},
 		{name: "grayGlacierBlock", block: c.GrayGlacierBlock, optional: true},
 		{name: "mergeNetsplitBlock", block: c.MergeNetsplitBlock, optional: true},
-		{name: "shanghaiBlock", block: c.ShanghaiBlock, optional: true},
 		{name: "cancunBlock", block: c.CancunBlock, optional: true},
+		{name: "shanghaiTime", timestamp: c.ShanghaiTime},
```
Comment (rjl493456442) [inline params/config.go:633]: Let's define Cancun as time and move it after the Shanghai.

File: core/forkid/forkid.go [hunk 8]
```diff
@@ -129,7 +148,11 @@ func newFilter(config *params.ChainConfig, genesis common.Hash, headfn func() ui
 	// Add two sentries to simplify the fork checks and don't require special
 	// casing the last one.
 	forks = append(forks, math.MaxUint64) // Last fork will never be passed
-
+	if len(forksByTime) == 0 {
+		forksByBlock = append(forks, math.MaxUint64) // Last fork will never be passed
+	} else {
+		forksByTime = append(forks, math.MaxUint64) // Last fork will never be passed
```
[the remark this answers is not in the document]
Comment (rjl493456442) [reply core/forkid/forkid.go:154]: Looks like a mistake, should be something like this?
```go
	if len(forksByTime) == 0 {
		forksByBlock = append(forksByBlock, math.MaxUint64) // Last fork will never be passed
	} else {
		forksByTime = append(forksByTime, math.MaxUint64) // Last fork will never be passed
	}
```

File: core/forkid/forkid.go [hunk 9]
```diff
@@ -163,7 +191,7 @@ func newFilter(config *params.ChainConfig, genesis common.Hash, headfn func() ui
 			if sums[i] == id.Hash {
 				// Fork checksum matched, check if a remote future fork block already passed
 				// locally without the local node being aware of it (rule #1a).
-				if id.Next > 0 && head >= id.Next {
+				if id.Next > 0 && (head >= id.Next || (id.Next > timestampThreshold && time >= id.Next)) {
```
Comment (rjl493456442) [inline core/forkid/forkid.go:194]: the head is either block or time, we can just compare it with `id.Next`, isn't it?

like 

```go
if id.Next > 0 && head >= id.Next {
      ....
}
```

File: core/headerchain.go (hunk 4 above)
Comment (holiman) [reply core/headerchain.go:587]: > If headTime is zero, it must be block mode;
> If headTime is non-zero, it must be time mode;

If both are zero, it 's moot: we're going to the genesis. However: the genesis _number_ is `0`, but is it guaranteed that the genesis _time_ is zero? 

Regardless, I think the code _as is_ works (goes into numbermode at `0,0`), and so does @rjl493456442 's suggestion.

File: core/headerchain.go [hunk 10]
```diff
@@ -566,15 +566,33 @@ type (
...
 		origin     = true
 	)
-	for hdr := hc.CurrentHeader(); hdr != nil && hdr.Number.Uint64() > head; hdr = hc.CurrentHeader() {
+	done := func(header *types.Header) bool {
+		if headBlock != 0 || headTime == 0 {
+			return header.Number.Uint64() <= headBlock
+		}
+		return header.Time <= headTime
+	}
+	for hdr := hc.CurrentHeader(); hdr != nil && !done(hdr); hdr = hc.CurrentHeader() {
```
Comment (holiman) [inline core/headerchain.go:592]: What happens if 
- I tell it to reset to `time=1`, 
- Genesis has `time=2`

Will it just loop forever or will something `break` out?

File: core/headerchain.go (hunk 4 above)
Comment (rjl493456442) [reply core/headerchain.go:587]: If both are zero, it means we want to rewind to genesis, (still time is zero, block mode).

The only exception is we want to rewind by timestamp, but the target is 0(zero timestamp). 
I feel like it's meaningless to have this case handling.

File: core/forkid/forkid.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (karalabe) [reply core/forkid/forkid.go:154]: Ah shit, yeah, that's a bad copy paste. Fixed.

```
// Pick the head comparison based on fork progression
head := block
if i >= len(forksByBlock) {
	head = time
}
```

If there are no timestamp based forks defined yet and we're past the last block based fork, then without this extra append  we switch over into timestamp mode, whereas we should remain in block mode.

TL;DR: Test #14 fails without the append :D 

Appending to forksByTime is not used, I just made it to be consistent if we change something in the future.

File: params/config.go (hunk 6 above)
Comment (karalabe) [reply params/config.go:374]: Guillaume's Verkle PR is based on the Cancun fork and block numbers. I didn't want to risk breaking that here. We can switch Cancun in a followup PR (I need it for 4844 anyway) and define Osaka too

File: params/config.go (hunk 7 above)
Comment (karalabe) [reply params/config.go:633]: I'll do it in a followup PR, it will be a bit breaking and don't want to convolute it with this change.

File: core/forkid/forkid.go (hunk 9 above)
Comment (karalabe) [reply core/forkid/forkid.go:194]: That breaks test 27 :)

File: core/headerchain.go (hunk 10 above)
Comment (karalabe) [reply core/headerchain.go:592]: Very good question :trollface:

File: core/forkid/forkid.go (hunk 9 above)
Comment (rjl493456442) [reply core/forkid/forkid.go:194]: Yeah, I get it.

`head` can represent it's a block or a timestamp of *local chain*, it's deterministic. But `id.Next` can refer to a block or a timestamp of *remote chain*, it's unknown. 

E.g. local chain defines all block-based forks without timestamp-based forks. Remote defines the same block-based forks + a timestamp fork as Next.

So we need this extra condition `id.Next > timestampThreshold ` to determine if it's a timestamp or a block number.

**Review by rjl493456442 — APPROVED** [review]
Mostly looks good to me!

File: core/headerchain.go (hunk 4 above)
Comment (karalabe) [reply core/headerchain.go:587]: I changed it to

```go
	done := func(header *types.Header) bool {
		if headTime > 0 {
			return header.Time <= headTime
		}
		return header.Number.Uint64() <= headBlock
	}
```

Seems cleaner to reason about but it's the exact same code as Gary's suggestion

Comment (karalabe) [reply core/headerchain.go:587]: It's also analogous to the rest of the comparisons across the codebase.

File: core/headerchain.go (hunk 5 above)
Comment (karalabe) [reply core/headerchain.go:614]: No, we always either want to revert based on timestamps or blocks. What this statement does it that it switches the timestamp based sethead to block based (since we already passed the timestamp).

Comment (karalabe) [reply core/headerchain.go:614]: Added a comment to highlight this explicit thought

File: core/headerchain.go (hunk 10 above)
Comment (karalabe) [reply core/headerchain.go:592]: Good catch ;)

Comment (karalabe) [reply core/headerchain.go:592]: I guess we should error out in this case, since there's no meaningful solution to rolling back the genesis

Comment (karalabe) [reply core/headerchain.go:605]: Fixed, also added shanghai overrides and fork console banner fixes to make this case actually testable.

Comment (karalabe) [reply core/headerchain.go:605]: PTAL
